Watchman.Discord/Areas/Initialization/Services/ServerScanningService.cs

```diff
@@ -85,10 +90,9 @@ private IEnumerable<DomainModel.Messages.Message> ConvertToMessages(IEnumerable<
         private List<Message> ReadMessages(DiscordServerContext server, ChannelContext channel, int limit, ulong lastMessageId = 0)
         {
             var messages = lastMessageId == 0
-                ? this._messagesHistoryService.ReadMessagesAsync(server, channel, limit).Result.ToList()
-                : this._messagesHistoryService.ReadMessagesAsync(server, channel, limit, lastMessageId, goBefore: true).Result.ToList();
-
-            return messages;
+                ? this._messagesHistoryService.ReadMessagesAsync(server, channel, limit).Result
+                : this._messagesHistoryService.ReadMessagesAsync(server, channel, limit, lastMessageId, goBefore: true).Result;
+            return messages.Where(x => x.Contexts.User.Id != this.BotContext.Id).ToList();
         }
```

Comment:
1. Ta metoda powinna IEnumerable<Message> zwraca i pozbądź się ToList bo po co to ma być?
2. Wcześniej w tej samej klasie masz Count == 0 zamień na !Any
3. Ogólnie chodzi o to abyś nie używał ToList i żebyś miał IEneumerable w tej klasie.

Reply:
a tak w ogóle po co te blokowanie? Zrób awaity i później w metodzie wywołującej do ReadMessages dodaj też await.

Reply:
1. I tak prędzej czy później muszę użyć ToList przed sprawdzeniem za pomocą .Any() czy są jakieś elementy, aby podwójnie nie pobierać wiadomości
2. to zmienię
3. Ta metoda i tak jest prywatna, więc mało to zmienia czy zwraca List czy IEnumerable. Mogę ewentualnie zmienić na IList albo IReadOnlyCollection

Reply:
Ok rozumiem, ale pamiętaj aby dać te awaity.

Poza tym mam pytanie, bo dopiero teraz poznałem pojęcie multiple enumeration i te operatory Linq są obliczane przy foreach ale czy też są obliczane dla każdego wywołania Any, Count, First itd. ?

Reply:
jeśli wyrażenia linq są wywołane na IEnumerable to tak, za każdym razem jest od nowa obliczane
jeśli jest później na taki IEnumerable zamieniony na jakąś kolekcję bez lazy loading (np. Lista albo Array) to wtedy nie jest już liczone ponownie

Reply:
czyli że jak nie dam ToList/ToArray i potem wywołam na tym np: First i później np: Any następnie zrobię foreach'a do tej kolekcji to wtedy jest obliczane trzy razy? (Dla First, Any i foreach kolejno?)

Reply:
tak, tylko dla First oraz Any zostanie wyliczony jedynie pierwszy element
więc tutaj zależy jak złożona jest ta kolekcja (jeśli ma np. pogrupować to i tak długo zajmie pobranie tego jednego elementu) ale nie wiem jak to dokładnie jest